rpc/query.go

```diff
@@ -777,6 +777,48 @@ func (t *ThetaRPCService) GetEenpByHeight(args *GetEenpByHeightArgs, result *Get
 	return nil
 }
 
+// ------------------------------ GetEenpStake -----------------------------------
+
+type GetEenpStakeByHeightArgs struct {
+	Height        common.JSONUint64 `json:"height"`
+	Source        common.Address    `json:"source"`
+	Holder        common.Address    `json:"holder"`
+	WithdrawnOnly bool              `json:"withdrawn_only"`
+}
+
+type GetEenpStakeResult struct {
+	Stake core.Stake `json:"stake"`
+}
+
+func (t *ThetaRPCService) GetEenpStakeByHeight(args *GetEenpStakeByHeightArgs, result *GetEenpStakeResult) (err error) {
+	deliveredView, err := t.ledger.GetDeliveredSnapshot()
+	if err != nil {
+		return err
+	}
+
+	db := deliveredView.GetDB()
+	height := uint64(args.Height)
+
+	var stake *core.Stake
+	blocks := t.chain.FindBlocksByHeight(height)
+	for _, b := range blocks {
+		stateRoot := b.StateHash
```

Comment:
We should check block.Status to find the best block. We had bug caused by similar code before in the case of chain having more than 1 branch. 

We can do something like this:
```
		var candidate *core.ExtendedBlock
		for _, b := range blocks {
			switch b.Status {
			case core.BlockStatusDirectlyFinalized, core.BlockStatusIndirectlyFinalized, core.BlockStatusTrusted:
				candidate = b
				break
			case core.BlockStatusCommitted, core.BlockStatusValid, core.BlockStatusPending:
				if candidate == nil || b.Status > candidate.Status {
					candidate = b
				}
			}
		}
		if candidate == nil {
			...
		}
```

(Probably this should be a method in core/chain.go)

Reply:
i see. i actually copied this code from existing "GetEenpByHeight" (line 759). shall we change both places?

Reply:
Yeah, let's refactor this into a util function so it can be called by both `GetEenpByHeight()` and `GetEenpStakeByHeight()`.